src/docmanager/index.ts

```diff
@@ -0,0 +1,955 @@
+      if ((ext in exts) || ('.*' in exts)) {
+        factories.push(name);
+      }
+    }
+    // Add the default widget if it was not already added.
+    name = this._defaultWidgetFactory;
+    if (name && factories.indexOf(name) === -1) {
+      options = this._widgetFactories[name];
+      if (options.modelName in this._modelFactories) {
+        factories.push(name);
+      }
+    }
+    return factories;
+  }
+
+  /**
+   * Get the kernel preference.
+   */
+  getKernelPreference(path: string, widgetName: string): IKernelPreference {
+    let widgetFactoryEx = this._getWidgetFactoryEx(widgetName);
+    let modelFactoryEx = this._getModelFactoryEx(widgetName);
+    let language = modelFactoryEx.factory.preferredLanguage(path);
+    return {
+      language,
+      preferKernel: widgetFactoryEx.preferKernel,
+      canStartKernel: widgetFactoryEx.canStartKernel
+    };
+  }
+
+  /**
+   * Open a file and return the widget used to display the contents.
+   *
+   * @param path - The file path to open.
+   *
+   * @param widgetName - The name of the widget factory to use.
+   *
+   * @param kernel - An optional kernel name/id to override the default.
+   */
+  open(path: string, widgetName='default', kernel?: IKernelId): Widget {
+    let widget = new Widget();
+    let manager = this._contentsManager;
+    if (widgetName === 'default') {
+      widgetName = this.listWidgetFactories(path)[0];
+    }
+    let mFactoryEx = this._getModelFactoryEx(widgetName);
+    let lang = mFactoryEx.factory.preferredLanguage(path);
+    let id = this._contextManager.findContext(path, mFactoryEx.name);
+    if (id) {
+      this._createWidget(id, widgetName, widget, kernel);
+      return widget;
+    }
+    let model = mFactoryEx.factory.createNew(lang);
+    let opts = mFactoryEx.contentsOptions;
+    manager.get(path, opts).then(contents => {
+      if (contents.format === 'json') {
+        model.fromJSON(contents.content);
+      } else {
+        model.fromString(contents.content);
+      }
+      model.dirty = false;
+      id = this._createContext(path, model, widgetName, contents);
+      this._createWidget(id, widgetName, widget, kernel);
+    });
+    return widget;
+  }
+
+  /**
+   * Create a new file of the given name.
+   *
+   * @param path - The file path to use.
+   *
+   * @param widgetName - The name of the widget factory to use.
+   *
+   * @param kernel - An optional kernel name/id to override the default.
+   */
+  createNew(path: string, widgetName='default', kernel?: IKernelId): Widget {
+    let widget = new Widget();
+    let manager = this._contentsManager;
+    if (widgetName === 'default') {
+      widgetName = this.listWidgetFactories(path)[0];
+    }
+    let mFactoryEx = this._getModelFactoryEx(widgetName);
+    if (!mFactoryEx) {
+      return;
+    }
+    let lang = mFactoryEx.factory.preferredLanguage(path);
+    let model = mFactoryEx.factory.createNew(lang);
+    let opts = mFactoryEx.contentsOptions;
```

Comment:
We want to copy the contents here, right? Since we are setting `opts.content` below...